libc/math/exp.c

```diff
@@ -0,0 +1,255 @@
+/*
+ * Phoenix-RTOS
+ *
+ * POSIX.1-2017 standard library functions tests
+ *
+ * HEADER:
+ *    - math.h
+ *
+ * TESTED:
+ *    - exp(), frexp(), ldexp()
+ *    - log(), log2(), log10()
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Adam Debek
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <math.h>
+#include <limits.h>
+#include <float.h>
+#include <errno.h>
+#include <stdlib.h>
+
+#include "common.h"
+#include <unity_fixture.h>
+
+TEST_GROUP(math_exp);
+
+
+TEST_SETUP(math_exp)
+{
+}
+
+
+TEST_TEAR_DOWN(math_exp)
+{
+}
+
+
+TEST(math_exp, exp_basic)
+{
+	int i, iterrs = 100 * ITERR_FACTOR;
+	double acceptLoss = 50.0;
+	double xmax = log(DBL_MAX);
+	double xmin = 1.0;
+	double x, y, ymin, ymax, f, g, digLost;
+
+	for (i = 0; i < iterrs; i++) {
+		x = test_getRandomLog(xmin, xmax);
+		if (x < 0.0) {
+			ymin = -(xmax + x);
+			ymax = xmax;
+		}
+		else {
+			ymin = -xmax;
+			ymax = xmax - x;
+		}
+
+		y = (double)rand() / RAND_MAX * (ymax - ymin) + ymin;
+		f = exp(x) * exp(y);
+		g = exp(x + y);
+
+		digLost = test_checkResult(f, g);
+		test_check_digLost2("exp", x, y, digLost, acceptLoss);
+	}
+}
+
+
+TEST(math_exp, exp_special_cond)
+{
+}
+
+
+TEST(math_exp, frexp_basic)
+{
+	int i, e, iterrs = 10 * ITERR_FACTOR;
+	double acceptLoss = 1.0;
+	double max = DBL_MAX / 2.0;
+	double min = DBL_MIN;
+	double x, f, g, y, digLost;
+
+	for (i = 0; i < iterrs; i++) {
+		x = test_getRandomLog(min, max);
+
+		if (i % 2) {
+			x = -x;
+		}
+
+		y = frexp(x, &e);
+
+		if (fabs(y) >= 1.0 || fabs(y) < 0.5) {
+			char errStr[100];
+			sprintf(errStr, "frexp(%g, int *exp) returned %g - value out of range <0.5, 1)", x, y);
+			TEST_FAIL_MESSAGE(errStr);
+		}
+
+		if (e > 0.0) {
+			f = y * pow(2.0, e);
+		}
+		else {
+			f = y / pow(2.0, -e);
+		}
```

Comment:
```suggestion
		f = y / pow(2,0, e);
```

Reply:
I think you meant `f = y * pow(2.0, e)` but i would stick to as it is, cause `pow(2.0, e)` for negative values is more flat and could return less precise values.